sync/statesync/code_queue.go

```diff
@@ -0,0 +1,260 @@
+	// already-present code. Persisting all enables consumer-side retry.
+	// Note: markers are keyed by code hash, so repeated persists overwrite the same
+	// key rather than growing DB usage. The consumer deletes the marker after
+	// fulfilling the request (or when it detects code is already present).
+	selected := make([]common.Hash, 0, len(codeHashes))
+	for _, codeHash := range codeHashes {
+		customrawdb.AddCodeToFetch(batch, codeHash)
+		selected = append(selected, codeHash)
+	}
+
+	if err := batch.Write(); err != nil {
+		return fmt.Errorf("failed to write batch of code to fetch markers due to: %w", err)
+	}
+
+	q.enqueueWG.Add(1)
+	defer q.enqueueWG.Done()
+	for _, hash := range selected {
+		select {
+		case q.out <- hash:
+		case <-q.quit:
+			return errFailedToAddCodeHashesToQueue
+		}
+	}
+	return nil
+}
+
+// Finalize implements [syncpkg.CodeRequestQueue] by signaling no further code hashes will be added.
+// Waits for in-flight enqueues to complete, then closes the output channel.
+// If the queue was already closed due to early quit, returns errFailedToFinalizeCodeQueue.
+func (q *CodeQueue) Finalize() error {
+	q.enqueueWG.Wait()
+	// If already closed due to quit, report early-exit error.
+	if q.closed.get() == closeStateQuit {
+		return errFailedToFinalizeCodeQueue
+	}
+	// Mark as finalized and close the channel (no-op if already finalized).
+	q.closed.markFinalizedAndClose(q.out)
+	return nil
+}
+
+// recoverUnfetchedCodeHashes cleans out any codeToFetch markers from the database that are no longer
+// needed and returns any outstanding markers to the queue.
+func recoverUnfetchedCodeHashes(db ethdb.Database) ([]common.Hash, error) {
+	it := customrawdb.NewCodeToFetchIterator(db)
+	defer it.Release()
+
+	batch := db.NewBatch()
+	var codeHashes []common.Hash
+
+	for it.Next() {
+		codeHash := common.BytesToHash(it.Key()[len(customrawdb.CodeToFetchPrefix):])
+
+		// If we already have the codeHash, delete the marker from the database and continue.
+		if !rawdb.HasCode(db, codeHash) {
+			codeHashes = append(codeHashes, codeHash)
+			continue
+		}
+
+		customrawdb.DeleteCodeToFetch(batch, codeHash)
+		if batch.ValueSize() < ethdb.IdealBatchSize {
+			continue
+		}
+
+		// Write the batch to disk if it has reached the ideal batch size.
+		if err := batch.Write(); err != nil {
+			return nil, fmt.Errorf("failed to write batch removing old code markers: %w", err)
+		}
+		batch.Reset()
+	}
+
+	if err := it.Error(); err != nil {
+		return nil, fmt.Errorf("failed to iterate code entries to fetch: %w", err)
+	}
+
+	if batch.ValueSize() > 0 {
+		if err := batch.Write(); err != nil {
+			return nil, fmt.Errorf("failed to write batch removing old code markers: %w", err)
+		}
+	}
+
+	return codeHashes, nil
+}
+
+// atomicCloseState provides a tiny typed wrapper around an atomic uint32.
+// It exposes enum-like operations to make intent explicit at call sites.
+type atomicCloseState struct {
```

Comment:
(info) The fact that this is necessary points to significant problems in the consuming side of the queue, inherited from earlier work (i.e. not part of this PR). Fixing them is beyond the scope of the PR but I'm noting it nonetheless so we know to address it later.

(task) Please add an issue with a tech-debt label to fix the rest of this package.